vpr/src/route/route_tree_timing.cpp

```diff
@@ -85,397 +55,193 @@ static bool equal_approx(float a, float b) {
-    rt_node->u.child_list = linked_rt_edge;
-
-    downstream_rt_node->parent_node = rt_node;
-    downstream_rt_node->parent_switch = iswitch;
-
-    //Expand (recursively) each of the main-branch nodes adding any
-    //non-configurably connected nodes
-    //Sink is not included, so no need to pass in the node's ipin value.
-    for (int rr_node : main_branch_visited) {
-        add_non_configurable_to_route_tree(rr_node, false, all_visited, is_flat);
+    // Expand (recursively) each of the main-branch nodes adding any
+    // non-configurably connected nodes
+    // Sink is not included, so no need to pass in the node's ipin value.
+    for (RRNodeId rr_node : main_branch_visited) {
+        add_non_configurable_to_route_tree(tree.rr_node_to_rt_node.at(rr_node).value(), false, all_visited, is_flat);
     }
 
-    *sink_rt_node_ptr = sink_rt_node;
-    return (downstream_rt_node);
+    /* the first and last nodes we added.
+     * vec[size-1] works, because new_branch_inodes is guaranteed to contain at least [sink, found_node] */
+    vtr::optional<RouteTreeNode&> downstream_rt_node = tree.rr_node_to_rt_node.at(new_branch_inodes[new_branch_inodes.size() - 1]);
+    vtr::optional<RouteTreeNode&> sink_rt_node = tree.rr_node_to_rt_node.at(new_branch_inodes[0]);
+
+    return {downstream_rt_node, sink_rt_node};
 }
 
-static t_rt_node* add_non_configurable_to_route_tree(const int rr_node,
-                                                     const bool reached_by_non_configurable_edge,
-                                                     std::unordered_set<int>& visited,
-                                                     bool is_flat) {
-    t_rt_node* rt_node = nullptr;
+static void add_non_configurable_to_route_tree(RouteTreeNode& rt_node,
+                                               const bool reached_by_non_configurable_edge,
+                                               std::unordered_set<RRNodeId>& visited,
+                                               bool is_flat) {
+    if (visited.count(rt_node.inode) && reached_by_non_configurable_edge)
+        return;
+
+    visited.insert(rt_node.inode);
+
+    auto& device_ctx = g_vpr_ctx.device();
+    const auto& rr_graph = device_ctx.rr_graph;
 
-    if (!visited.count(rr_node) || !reached_by_non_configurable_edge) {
-        visited.insert(rr_node);
+    VTR_ASSERT(rt_node.root);
+    auto& rr_node_to_rt_node = rt_node.root.value().rr_node_to_rt_node;
+    const RRNodeId rr_node = rt_node.inode;
 
-        auto& device_ctx = g_vpr_ctx.device();
-        const auto& rr_graph = device_ctx.rr_graph;
+    for (int iedge : rr_graph.non_configurable_edges(rr_node)) {
+        // Recursive case: expand children
+        VTR_ASSERT(!rr_graph.edge_is_configurable(rr_node, iedge));
+        RRNodeId to_rr_node = rr_graph.edge_sink_node(rr_node, iedge);
 
-        rt_node = rr_node_to_rt_node[rr_node];
+        if (rr_node_to_rt_node.count(to_rr_node)) // TODO: not 100% sure about this
```

Comment:
Since the for loop calling this function is iterating over an unordered_set, you cannot be sure of the order of visiting. I think having this 'if' block makes sense. However, wouldn't it be better to iterate over the nodes in the normal order of the tree (from root to leaf) in the function that calls this function? Then you can change this 'if' condition to something like an assertion: 'VTR_ASSERT(rr_node_to_rt_node.count(to_rr_node) == 0).'

Reply:
I think that's not worth the trouble of debugging nonconf branch creation again :P